official/datasets/movielens.py

```diff
@@ -0,0 +1,308 @@
+
+MAX_RATING = 5
+
+
+NUM_RATINGS = {
+    ML_1M: 1000209,
+    ML_20M: 20000263
+}
+
+
+def _download_and_clean(dataset, data_dir):
+  """Download MovieLens dataset in a standard format.
+
+  This function downloads the specified MovieLens format and coerces it into a
+  standard format. The only difference between the ml-1m and ml-20m datasets
+  after this point (other than size, of course) is that the 1m dataset uses
+  whole number ratings while the 20m dataset allows half integer ratings.
+  """
+  if dataset not in DATASETS:
+    raise ValueError("dataset must be one of: {}".format(",".join(DATASETS)))
+
+  data_subdir = os.path.join(data_dir, dataset)
+
+  expected_files = ["{}.zip".format(dataset), RATINGS_FILE, MOVIES_FILE]
+
+  tf.gfile.MakeDirs(data_subdir)
+  if set(expected_files).intersection(
+      tf.gfile.ListDirectory(data_subdir)) == set(expected_files):
+    tf.logging.info("Dataset {} has already been downloaded".format(dataset))
+    return
+
+  url = "{}{}.zip".format(_DATA_URL, dataset)
+
+  temp_dir = tempfile.mkdtemp()
+  try:
+    zip_path = os.path.join(temp_dir, "{}.zip".format(dataset))
+    def _progress(count, block_size, total_size):
+      sys.stdout.write("\r>> Downloading {} {:.1f}%".format(
+          zip_path, 100.0 * count * block_size / total_size))
+      sys.stdout.flush()
+
+    zip_path, _ = urllib.request.urlretrieve(url, zip_path, _progress)
+    statinfo = os.stat(zip_path)
+    # A new line to clear the carriage return from download progress
+    # tf.logging.info is not applicable here
+    print()
+    tf.logging.info(
+        "Successfully downloaded {} {} bytes".format(
+            zip_path, statinfo.st_size))
+
+    zipfile.ZipFile(zip_path, "r").extractall(temp_dir)
+
+    if dataset == ML_1M:
+      _regularize_1m_dataset(temp_dir)
+    else:
+      _regularize_20m_dataset(temp_dir)
+
+    for fname in tf.gfile.ListDirectory(temp_dir):
+      tf.gfile.Copy(os.path.join(temp_dir, fname),
+                    os.path.join(data_subdir, fname))
+
+  finally:
+    tf.gfile.DeleteRecursively(temp_dir)
+
+
+def _write_csv(input_path, output_path, names, skip_first, separator=","):
+  """Transform csv to a regularized format.
+
+  Args:
+    input_path: The path of the raw csv.
+    output_path: The path of the cleaned csv.
+    names: The csv column names.
+    skip_first: Boolean of whether to skip the first line of the raw csv.
+    separator: Character used to separate fields in the raw csv.
+  """
+  if six.PY2:
+    names = [n.decode("utf-u") for n in names]
+
+  with tf.gfile.Open(output_path, "wb") as f_out, \
+      tf.gfile.Open(input_path, "rb") as f_in:
+
+    # Write column names to the csv.
+    f_out.write(",".join(names).encode("utf-8"))
+    f_out.write(b"\n")
+    for i, line in enumerate(f_in):
+      if i == 0 and skip_first:
+        continue  # ignore existing labels in the csv
+
+      line = line.decode("utf-8", errors="ignore")
+      fields = line.split(separator)
```

Comment:
Just want to make sure that we process the data with ',' as separator correctly. Usually data with ',' in content will quote the comma, and in this case, we didn't handle it correctly.

Reply:
Yes. One of the two datasets uses commas as a separator, and has already quoted the fields with commas in the literal text. This is there to avoid double quoting those entries.